File: leet_code/linked_lists/find_intersection.rb
def find_intersection(head_a, head_b)
  return nil if !head_a || !head_b
  return head_a if head_a == head_b

  list_one = head_a
  list_two = head_b

  while list_one && list_two && (list_one != list_two)
    list_one = list_one.next
    list_two = list_two.next

    return list_one if list_one == list_two
    list_one = head_b if !list_one
    list_two = head_a if !list_two
  end
  list_one
end

File: leet_code/linked_lists/find_intersection.js
const findIntersection = (headA, headB) => {
  if (!headA || !headB) return null;
  else if (headA == headB) return headA;

  let listOne = headA;
  let listTwo = headB;

  while (listOne && listTwo && (listOne != listTwo)) {
    listOne = listOne.next;
    listTwo = listTwo.next;

    if (listOne == listTwo) return listOne;
    if (!listOne) listOne = headB;
    if (!listTwo) listTwo = headA;
  }
  return listOne;
};
